design/user_controller_workspaces.md

```diff
@@ -0,0 +1,143 @@
+ That's it. It doesn't keep track of it (more on that later).
+
+Writes just append the path to be written to the workspace path: `${workspaceAddr}/${writePath}`
+Deletes are also writes, but of a sentinel tombstone value.
+Reads are now a getObject/listObjects call on that prefix, implementing our "DiffIterator".
+
+When calling `commit`, we take that diffIterator and apply it. Once we have a commit ID, we do a `CAS` on the ref: if no one else committed so far, this completes the commit process successfully. Otherwise, its an `ErrConcurrentCommit` and should be retried/aborted.
+Reading these refs is done using the same batching method we currently do for PG, so we could support pretty much ANY external storage for this as long as it is:
+
+* Read-after-write consistent
+* Has at least one of `{CAS, conditional-write, serializable transaction}`
+
+On this list are: DynamoDB (which we can abstract away(!)), PostgreSQL, MySQL, MongoDB, even a Hive Metastore..
+Using good object stores (i.e., practically anything other than S3), this could be done with no DB at all, with some additional latency.
+
+It is the user's responsibility to ensure they commit at the end of their job/pipeline/script,
+but it's possible to pass around that `workspaceID` between jobs and processes, even if reading/writing concurrently.
+
+Once a commit is done for a given workspace, it should no longer be used. if using the lakeFS SDK (as shown above),
+calling any operation on a `commit()`ed `Workspace` object will raise an error. We can go further and place a "lockfile" (`_lakefs_committed/${workspaceID}`) and check it before writing if we think this is neccesary (on a best effort basis).
+
+## Using the S3 gateway
+
+just append a workspace ID: `s3://my-repo/master+abc123/my/collection/` to read/write.
+otherwise, just pass a ref as you normally would, to read committed data.
+
+## Out-of-band ingest
+
+Now that a commit is basically a diffIterator on top of `listObjects` we can do something similar for stream processing.
+Working with streams is different: we can't always ensure no writes are happening, but fortunately, most streaming processors 
+do [guarantee immutability](https://stackoverflow.com/a/61201083): A written file is never overwritten, only written once and potentially deleted: this allows us to take
+a regular old object store path and do:
+
+```sh
+$ lakectl ingest commit \
+    --from-path "s3://bucket/kafka/events/2021/" \
+    --to-path "lakefs://my-repo@ingest-branch/kafka/events/2021/" \
+    -m "ingest events: $(date -u)"
+
+# Or just see what changes would be introduced
+$ lakectl ingest diff \
+    --from-path "s3://bucket/kafka/events/2021/" \
+    --to-path "lakefs://my-repo@ingest-branch/kafka/events/2021/" \
+```
+Here, instead of treating the given object store location as a diffIterator, we look at it as a subtree, to be diffed with the underlying commit.
```

Comment:
This ingest can be very expensive, because it involves a listing.  It makes ops tricky: it is never clear _which_ commit received a file.  Indeed, _commits receive files out of order with a unboundedly higher skew_ if the ingester is multi-threaded.  _Any_ skew, no matter how small, gets amplified to the _time between ingests_.

## Example

We ingest files every 2 minutes.

1. Commit starts for all files starting `PREFIX-12:34`.
2. Commit lists first page of files, up to `PREFIX-12:34:56.002`.
3. Another ingester thread finally manages to write `PREFIX-12:34:56.001`.
4. Committed continues listing, up to `PREFIX-12:35:59.999`.
5. File `PREFIX-12:34:56.001` is missing.  It arrived a few milliseconds too late, and shows up in a commit two minutes later.

Reply:
# On the bright side...

It's not _really_ "unboundedly higher skew", it's "skewed by time between ingests".

# On the cloud side...

Such scenarios can break semantics.  E.g. `_success` and similar files can easily be written into a commit that doesn't include all their files, or they might be named `_finished_writing_up_to_12:36` (and lie).

Reply:
Possible mitigation of the `_success` files scenario, going over all files under the prefix twice:
1. Pick up *only* `_success` files.
2. Pick up everything which isn't a `_success` file.